dotCMS/src/main/java/com/dotcms/filters/interceptor/saml/SamlWebUtils.java

```diff
@@ -185,4 +214,78 @@ protected HttpSession renewSession(final HttpServletRequest request, final HttpS
     }
 
 
+    /**
+     * Return the relay state for the SAML request if it is possible
+     * it could be a specific delegate class implemented or a generic one based on the velocity template indexed by the property auth.relaystate on the app SAML config
+     * for instance this config: auth.relaystate=companyCode=request.getParameter('companyCode')
+     * Will took the companyCode parameter from the request and will put it on the relay state
+     * Otherwise null will be returned
+     * @param request
+     * @param response
+     * @param identityProviderConfiguration
+     * @param siteIdentifier
+     * @return
+     */
+    public String getRelayState(final HttpServletRequest request,
+                                final HttpServletResponse response,
+                                final IdentityProviderConfiguration identityProviderConfiguration,
+                                final String siteIdentifier) {
+
+        final String relayState = identityProviderConfiguration.containsOptionalProperty(AUTH_RELAYSTATE_KEY)?
+                identityProviderConfiguration.getOptionalProperty(AUTH_RELAYSTATE_KEY).toString(): null;
+        return getRelayState(request, response, identityProviderConfiguration, relayState, siteIdentifier);
+    }
+
+    /**
+     * Based on the relayStateTemplate value will find a specific delegate class implemented or a generic one based on the velocity template
+     * for instance if the value for the parameter relayStateTemplate is: companyCode=request.getParameter('companyCode')
+     * Will took the companyCode parameter from the request and will put it on the relay state
+     * Otherwise null will be returned
+     * @param request
+     * @param response
+     * @param identityProviderConfiguration
+     * @param relayStateTemplate
+     * @param siteIdentifier
+     * @return
+     */
+    public String getRelayState(final HttpServletRequest request,
+                                final HttpServletResponse response,
+                                final IdentityProviderConfiguration identityProviderConfiguration,
+                                final String relayStateTemplate,
+                                final String siteIdentifier) {
+
+        request.setAttribute(AUTH_RELAYSTATE_KEY, relayStateTemplate);
+
+        final RelayStateStrategy relayStateStrategy = RELAY_STATE_STRATEGY_MAP.getOrDefault(siteIdentifier, DEFAULT_VELOCITY_RELAY_STATE_STRATEGY);
+
+        final String relayStateValue =  relayStateStrategy.apply(request, response, identityProviderConfiguration);
+
+        Optional.ofNullable(request.getSession(false)).ifPresent(session -> session.setAttribute(AUTH_RELAYSTATE_KEY, relayStateValue));
```

Comment:
Shouldn't we always set it here?  meaning, create the session if it does not exist at this point?  the user is in the middle of a login